File: src/com/jialtang/leetcode/002AddTwoNumbers.go
package main

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var root = &ListNode{}
	var tail = root

	var carry = 0
	for l1 != nil || l2 != nil {
		var node = &ListNode{}
		var val1 = 0
		var val2 = 0
		if l1 != nil {
			val1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			val2 = l2.Val
			l2 = l2.Next
		}
		node.Val = (val1 + val2 + carry) % 10
		carry = (val1 + val2 + carry) / 10

		tail.Next = node
		tail = tail.Next
	}

	if carry != 0 {
		node := newANode(carry)
		tail.Next = node
		tail = tail.Next
	}

	return root.Next
}

func newANode(val int) *ListNode {
	l := &ListNode{}
	l.Val = val
	return l
}

type ListNode struct {
	Val  int
	Next *ListNode
}

File: src/com/jialtang/leetcode/_002AddTwoNumbers.java
package com.jialtang.leetcode;

import com.jialtang.leetcode.common.ListNode;

public class _002AddTwoNumbers {

    public static void main(String[] args) {
        ListNode list1 = ListNode.Create.createNode(9);
        System.out.printf("List: [%s]", list1.toString());
        ListNode list2 = ListNode.Create.createNode(9);
        System.out.printf("List: [%s]", list2.toString());

        ListNode listNode = new Solution().addTwoNumbers(list1, list2);
        System.out.printf("List: [%s]", listNode.toString());
    }

    static class Solution {
        public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
            ListNode root = new ListNode();
            ListNode cur = root;

            ListNode s1 = l1;
            ListNode s2 = l2;
            int carr = 0;

            while (s1 != null || s2 != null) {
                int val1 = getVal(s1);
                int val2 = getVal(s2);

                ListNode newNode = new ListNode();
                newNode.val = (val1 + val2 + carr) % 10;
                carr = (val1 + val2 + carr) / 10;
                cur.next = newNode;
                cur = nextNode(cur);

                s1 = nextNode(s1);
                s2 = nextNode(s2);
            }

            if (carr != 0) {
                ListNode newNode = new ListNode();
                newNode.val = (carr) % 10;
                cur.next = newNode;
                cur = nextNode(cur);
            }

            return root.next;
        }

        private ListNode nextNode(ListNode node) {
            if (node == null) {
                return null;
            }

            return node.next;
        }

        private int getVal(ListNode node) {
            if (node == null) {
                return 0;
            }
            return node.val;
        }
    }
}
